Adds Koan Update and Reset commands

indented-automation opened the pull request:
﻿# PR Summary

Adds commands to update a users repository state.

* Reset-PSKoan
* Update-PSKoan

## Context

Allows updates to the users repository without adversely affecting progress.

Reset-PSKoan allows an individual koan to be reset to the module default. Koans are selected by name and, optionally, context.

Update-PSKoan performs the following activities:

* Replays completed koans into files copied from the module.
* Adds missing topics
* Fixes drifting file paths
* Removes topics no longer present in the module path

## Changes

New content only.

## Checklist

- [x] Pull Request has a meaningful title.
- [x] Summarised changes.
- [x] Pull Request is ready to merge & is not WIP.
- [x] Added tests / only testable interactively.
  - Make sure you add a new test if old tests do not effectively test the code changed.
- [x] Added documentation / opened issue to track adding documentation at a later date.

vexx32 commented on PSKoans/Public/Update-PSKoan.ps1:
```diff
@@ -0,0 +1,73 @@
+        Where-Object { -not $Topic -or $_.BaseName -replace '\.Koans$' -in $Topic } |
+        Group-Object { $_.BaseName -replace '\.Koans$' } -AsHashTable -AsString
+
+    $TopicList = [System.Collections.Generic.HashSet[String]]::new([System.StringComparer]::InvariantCultureIgnoreCase)
```
```suggestion
    $TopicList = [HashSet[String]]::new(
        [System.StringComparer]::InvariantCultureIgnoreCase)
```
Also, make sure you add the appropriate `using namespace` to the top of the file. 🙂

vexx32 commented on PSKoans/Public/Update-PSKoan.ps1:
```diff
@@ -0,0 +1,73 @@
+            }
+        }
+    }
+}
```
Newline at EOF please! :)

indented-automation replied:
Fixed all :)

vexx32 commented on Tests/Functions/Private/Update-PSKoanFile.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+                    ModuleFilePath = Join-Path $TestDrive 'Module\AboutSomething.ps1'
+                }
+            }
+            New-Item 'TestDrive:\Koans' -ItemType Directory
```
Probably best to just use `TestDrive:Koans` or else make use of `[IO.Path]::DirectorySeparatorChar`

indented-automation replied:
Removed drive use

vexx32 commented on Tests/Functions/Public/Reset-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,94 @@
+                }
+            }
+            New-Item 'TestDrive:\Koans' -ItemType Directory
+            New-Item 'TestDrive:\Module' -ItemType Directory
```
Here too

indented-automation replied:
Removed drive use

vexx32 commented on Tests/Functions/Public/Reset-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,94 @@
+            Mock Get-PSKoanFilePath {
+                [PSCustomObject]@{
+                    UserFilePath   = Join-Path $TestDrive 'Koans\AboutSomething.ps1'
+                    ModuleFilePath = Join-Path $TestDrive 'Module\AboutSomething.ps1'
```
Here too

indented-automation replied:
This one is fine, Join-Path handles fixing up the path separator for each platform.

vexx32 commented on Tests/Functions/Private/Update-PSKoanFile.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+            Mock Get-PSKoanFilePath {
+                [PSCustomObject]@{
+                    UserFilePath   = Join-Path $TestDrive 'Koans\AboutSomething.ps1'
+                    ModuleFilePath = Join-Path $TestDrive 'Module\AboutSomething.ps1'
```
Oops, here too

indented-automation replied:
Join-Path again :)

vexx32 commented on Tests/Functions/Public/Reset-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,94 @@
+            Reset-PSKoan -Topic AboutSomething -Name 'reset content' -Confirm:$false
+
+            (Get-PSKoanFilePath).UserFilePath | Should -FileContentMatch '1 | Should -Be 1'
+            (Get-PSKoanFilePath).UserFilePath | Should -FileContentMatch '__ | Should -Be 2'
```
This is used repeatedly -- should this be defined as a variable in a BeforeAll?

indented-automation replied:
Fixed.

vexx32 commented on Tests/Functions/Public/Update-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+                Mock Get-ChildItem -ParameterFilter { $LiteralPath -match 'PSKoans' } -MockWith {
+                    [System.IO.FileInfo]'Module\Koans\Missing.Koans.ps1'
+                    [System.IO.FileInfo]'Module\Koans\Right\Path.Koans.ps1'
+                    [System.IO.FileInfo]'Module\Koans\Update.Koans.ps1'
```
These directory separators too

indented-automation replied:
Fixed. Replaced with System.IO.Path.Combine.

vexx32 commented on Tests/Functions/Public/Update-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+                Mock Get-ChildItem -ParameterFilter { $LiteralPath -match 'TestDrive'} -MockWith {
+                    [System.IO.FileInfo]"${TestDrive}\Koans\Wrong\Path.Koans.ps1"
+                    [System.IO.FileInfo]"${TestDrive}\Koans\Update.Koans.ps1"
+                    [System.IO.FileInfo]"${TestDrive}\Koans\Remove.Koans.ps1"
```
And these

indented-automation replied:
Replaced with System.IO.Path.Combine.

vexx32 commented on Tests/Functions/Public/Update-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+            }
+
+            It 'should move incorrectly placed topics' {
+                $directory = New-Item (Join-Path $TestDrive 'Koans\Wrong') -ItemType Directory
```
here too

indented-automation replied:
Not a problem, Join-Path makes right.

vexx32 commented on Tests/Functions/Public/Update-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+            }
+
+            It 'should remove discarded topics' {
+                $oldTopicPath = Join-Path $TestDrive 'Koans\OldTopic.koans.ps1'
```
and here

indented-automation replied:
Join-Path fixes.

vexx32 commented on PSKoans/Private/Get-PSKoanFilePath.ps1:
```diff
@@ -0,0 +1,50 @@
+function Get-PSKoanFilePath {
+    <#
+    .SYNOPSIS
+        Get the paths for module and user koan topics.
```
You can ignore this. /facepalm

vexx32 commented:
@indented-automation friendly ping so you don't forget 😁

indented-automation commented:
Sorry sorry, a touch busy. Few more days / the weekend and I should have
time once more :)

On Tue, 20 Aug 2019 at 14:03, vexx32 <notifications@github.com> wrote:

> @indented-automation <https://github.com/indented-automation> friendly
> ping so you don't forget 😁
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vexx32/PSKoans/pull/223?email_source=notifications&email_token=ADB3UDXS3AMTTNVKYLGKKXDQFPTRRA5CNFSM4IKC2RM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4WGSZY#issuecomment-523004263>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADB3UDVCRUTKVLVRL4NGSXLQFPTRRANCNFSM4IKC2RMQ>
> .
>

vexx32 commented:
All good, whenever you have time! 😊  💖

chridentitv commented on PSKoans/Private/Get-PSKoanFilePath.ps1:
```diff
@@ -0,0 +1,50 @@
+function Get-PSKoanFilePath {
```
No problem. It has been renamed.

indented-automation replied:
Done.

chridentitv commented on PSKoans/Private/Get-PSKoanFilePath.ps1:
```diff
@@ -0,0 +1,50 @@
+    process {
+        Join-Path -Path $script:ModuleRoot -ChildPath 'Koans' |
+            Get-ChildItem -Recurse -Filter *.Koans.ps1 |
+            Where-Object { -not $Topic -or $_.BaseName -replace '\.Koans$' -in $Topic } |
```
Done.

indented-automation replied:
Done.

chridentitv commented on PSKoans/Private/Get-PSKoanIt.ps1:
```diff
@@ -0,0 +1,58 @@
+        },
+        $true
+    ) | ForEach-Object {
+        $contextName = ''
```
Done.

indented-automation replied:
Done.

indented-automation commented on PSKoans/Public/Update-PSKoan.ps1:
```diff
@@ -0,0 +1,76 @@
+
+                $ParentPath = Split-Path -Path $DestinationPath -Parent
+                if (-not (Test-Path -Path $ParentPath)) {
+                    $null = New-Item -Path $ParentPath -ItemType Directory
```
ew redirect ;) Done :)

indented-automation commented on Tests/Functions/Public/Reset-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,96 @@
+            New-Item (Join-Path $TestDrive 'Module') -ItemType Directory
+
+            Set-Content -Path (Get-PSKoanFilePath).ModuleFilePath -Value @(
+                "Describe 'AboutSomething' {"
```
Done.

indented-automation commented on Tests/Functions/Public/Reset-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,96 @@
+
+        BeforeEach {
+            Set-Content -Path (Get-PSKoanFilePath).UserFilePath -Value @(
+                "Describe 'AboutSomething' {"
```
Done.

indented-automation commented on Tests/Functions/Public/Update-PSKoan.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+            }
+
+            It 'should move incorrectly placed topics' {
+                $directory = New-Item (Join-Path $TestDrive 'Koans\Wrong') -ItemType Directory
```
Updated all instances of this.

chridentitv commented on PSKoans/Private/Get-PSKoanIt.ps1:
```diff
@@ -0,0 +1,58 @@
+            $node.GetCommandName() -eq 'It'
+        },
+        $true
+    ) | ForEach-Object {
```
Changed this one again, ForEach is *not* forgiving if there are no results. ForEach-Object was. Changed it to a foreach loop instead. Safer and still looses the pipeline here.

indented-automation replied:
Changed this one again, ForEach is not forgiving if there are no results. ForEach-Object was. Changed it to a foreach loop instead. Safer and still looses the pipeline here.

indented-automation commented:
I need to do a bit more testing. I would suggest that the Reset switch on Measure-Karma should be removed. Attempting to wire in the reset options here would make it unnecessarily complicated.

If you agree I'll do more testing to see if this can replace the existing Initialize and Reset functionality.

vexx32 commented:
Sounds good to me! :blush:

indented-automation commented on PSKoans/Private/Get-PSKoanIt.ps1:
```diff
@@ -0,0 +1,63 @@
+            {
+                param ( $node )
+
+                $node -is [System.Management.Automation.Language.CommandAst] -and
```
Sure, why not.

indented-automation commented on PSKoans/Private/Get-PSKoanIt.ps1:
```diff
@@ -0,0 +1,63 @@
+            if ($parentAst -is [System.Management.Automation.Language.CommandAst] -and $parentAst.GetCommandName() -eq 'Context') {
+                $contextName = $parentAst.CommandElements[1].Value
+            }
+            } until (-not $parentAst -or $contextName)
```
Done... not sure how I missed the indent there. Must have broken it when I decided to add process.

indented-automation commented on docs/Reset-PSKoan.md:
````diff
@@ -0,0 +1,75 @@
+```
+
+### CommonParameters
+This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable, -InformationAction, -InformationVariable, -OutVariable, -OutBuffer, -PipelineVariable, -Verbose, -WarningAction, and -WarningVariable. For more information, see [about_CommonParameters](http://go.microsoft.com/fwlink/?LinkID=113216).
````
Done.

indented-automation commented:
The fixes requested above plus:

* The name of the koan in Reset is now optional allowing a topic to be reset easily.
* Changed wildcard support in Reset-PSKoan to regex support (consistency).
* Added regex and array input for the Topic parameter to Reset-PSKoan.
* Added argument completers to the Update and Reset commands.
* Added more examples to the help file for Reset-PSKoan.
* Removed WhatIf and Confirm from the Measure-Karma document (syntax) since Reset has been removed.

indented-automation commented:
Wildcards, it's more welcoming and I think would be less surprising. I can swap everything over to wildcards if you think it appropriate.

vexx32 commented:
Also @indented-automation if you could do a quick rebase after you address the last comments so we can triple-check all your lovely tests that'd be greatly appreciated! 💖 😊

vexx32 commented on PSKoans/Private/Get-Koan.ps1:
```diff
@@ -9,7 +7,7 @@ function Get-Koan {
 
     .PARAMETER Topic
-        Speficy one or more topic names or patterns to filter the list. Regex matching is permitted.
+        Specify one or more topic names or patterns to filter the list. Regex matching is permitted.
```
```suggestion
        Specify one or more topic names or patterns to filter the list. Wildcards are permitted.
```

vexx32 replied:
Might also be worth adding `[SupportsWildcards()]` to the parameter as well. 🙂
